File: solutions/go/03/tree_dfs/main.go
package main

import "fmt"

func dfs(v, parent int, g [][]int, depth []int) {
	if parent == -1 {
		depth[v] = 0
	} else {
		depth[v] = depth[parent] + 1
	}
	for _, to := range g[v] {
		if to == parent {
			continue
		}
		dfs(to, v, g, depth)
	}
}

func main() {
	g := [][]int{{1, 2}, {0, 3}, {0}, {1}}
	depth := make([]int, 4)
	dfs(0, -1, g, depth)
	fmt.Println(depth)
}

File: solutions/python/03/tree_dfs/main.py
import sys

sys.setrecursionlimit(10**7)


def dfs(v: int, parent: int, g: list, depth: list) -> None:
    depth[v] = depth[parent] + 1 if parent != -1 else 0
    for to in g[v]:
        if to == parent:
            continue
        dfs(to, v, g, depth)


def main() -> None:
    g = [[1, 2], [0, 3], [0], [1]]
    depth = [0] * 4
    dfs(0, -1, g, depth)
    print(depth)


if __name__ == "__main__":
    main()
